Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,1047 @@
+
+Once we have added these tests to our test suite, we should re-run the entire test suite using `pytest tests` and check that our new Hypothesis-based tests are among the tests being run.
+</div>
+
+<!-- #endregion -->
+
+<!-- #region -->
+## Extending the Functionality of Strategies
+
+Hypothesis strategies can be enriched through the use of two methods: `.map()` and `.filter()`.
+These will permit us to leverage Hypothesis' core strategies to describe much more distinctive and diverse varieties of data.
+We also will see that there is a `st.data()` strategy, which will enable us to draw from strategies interactively from within our tests.
+
+
+### The `.map` method
+
+Hypothesis strategies have the `.map` method, which permits us to [perform a mapping on the data](https://hypothesis.readthedocs.io/en/latest/data.html#mapping) being produced by a strategy.
+This is achieved by passing the `.map` method a function (or any "callable");
+upon drawing a value from a strategy, Hypothesis will feed that value to the function held by `.map()`, and the strategy
+will return the value that was returned by the function.
+In this way the strategy's output is automatically "mapped" to a transformed value via the function that we provided.
+
+For example, if we want to draw only even-valued integers, we can simply use the following mapped strategy:
+
+```python
+# demonstrating the `.map()` method
+
+def make_even(x): 
+    return 2 * x
+
+# `even_integers` is now a strategy that will only return even
+# valued integers. This is achieved by ensuring that any integer
+# drawn by `st.integers()` is "mapped" to an even value
+# via the function x -> 2 * x
+even_integers = st.integers().map(make_even)
+```
+
+```python
+>>> even_integers.example()
+-15414
+```
+<!-- #endregion -->
+
+<!-- #region -->
+#### A Brief Aside: Lambda Expressions
+
+Python has a syntax, which we have yet to discuss, that permits us to conveniently define functions "on the fly".
+A "lambda expression" is a syntax for defining a simple one-line function, making that function available for use in-place.
+The key here is that, whereas standard functions first must be formally defined before they can be referenced in code, *a lambda expression can be written wherever a function is expected*.
+
+For example, we can simplify the above mapping example by defining our mapping-function _within_ the `.map()` method:
+
+```python
+# Using a lambda expression to define a function
+# "on the fly"
+even_integers = st.integers().map(lambda x: 2 * x)
+```
+
+```python
+>>> even_integers.example()
+220
+```
+
+In general, the syntax for defining a lambda expression is:
+
+```
+lambda <comma-separated variable names>: <expression using variables>
+```
+
+Note that lambda expressions are restricted compared to typical function definitions: they do not permit default values or keyword arguments in their signatures, and a lambda's "body" must fit on one line.
```

Comment:
Lambdas *do* permit default values and everything else you could put in a normal function signature; the restriction is that the function body must consist of a single expression - but that can span multiple physical lines!

Reply:
😮 I'm ashamed that I didn't know this!